org.eclipse.jdt.ls.tests/src/org/eclipse/jdt/ls/core/internal/handlers/CompletionHandlerTest.java

```diff
@@ -401,6 +399,131 @@ public void testCompletion_package() throws JavaModelException{
+
+		ICompilationUnit unit = getWorkingCopy(
+				"src/java/Foo.java",
+				"public class Foo {\n"
+						+ "    toStr"
+						+"}\n");
+		int[] loc = findCompletionLocation(unit, " toStr");
+		CompletionList list = server.completion(JsonMessageHelper.getParams(createCompletionRequest(unit, loc[0], loc[1]))).join().getRight();
+		assertNotNull(list);
+		List<CompletionItem> filtered = list.getItems().stream().filter((item)->{
+			return item.getDetail() != null && item.getDetail().startsWith("Override method in");
+		}).collect(Collectors.toList());
+		assertFalse("No override proposals", filtered.isEmpty());
+		CompletionItem oride = filtered.get(0);
+		assertEquals("toString", oride.getInsertText());
+		assertNull(oride.getTextEdit());
+		oride = server.resolveCompletionItem(oride).join();
+		assertNotNull(oride.getTextEdit());
+		String text = oride.getTextEdit().getNewText();
+		StringBuilder expectedText = new StringBuilder();
+		if (overridesSuperClass) {
+			expectedText.append("@Override\n");
+		}
+		expectedText.append("public String toString() {\n\t");
+		if (supportSnippets) {
+			expectedText.append("${0:");
+		}
+		expectedText.append("return super.toString();");
+		if (supportSnippets) {
+			expectedText.append("}");
+		}
+		expectedText.append("\n}");
+
+		assertEquals(expectedText.toString(), text);
+	}
+
+	private void testCompletion_interfaceMethodOverride(String projectName, boolean supportSnippets,
+			boolean overridesInterface) throws Exception {
+		if (project == null || !projectName.equals(project.getName())) {
+			importProjects("eclipse/" + projectName);
+			project = WorkspaceHelper.getProject(projectName);
+		}
+		mockClientPreferences(supportSnippets, true);
+
+		ICompilationUnit unit = getWorkingCopy(
+				"src/java/Foo.java",
+				"public class Foo implements Runnable{\n"
+						+ "    ru"
+						+"}\n");
+		int[] loc = findCompletionLocation(unit, " ru");
+		CompletionList list = server
+				.completion(JsonMessageHelper.getParams(createCompletionRequest(unit, loc[0], loc[1]))).join()
+				.getRight();
+		assertNotNull(list);
+		List<CompletionItem> filtered = list.getItems().stream().filter((item) -> {
+			return item.getDetail() != null && item.getDetail().startsWith("Override method in");
+		}).collect(Collectors.toList());
+		assertFalse("No override proposals", filtered.isEmpty());
+		CompletionItem oride = filtered.get(0);
+		assertEquals("run", oride.getInsertText());
+		assertNull(oride.getTextEdit());
+		oride = server.resolveCompletionItem(oride).join();
+		assertNotNull(oride.getTextEdit());
+		String text = oride.getTextEdit().getNewText();
+		StringBuilder expectedText = new StringBuilder();
+		if (overridesInterface) {
+			expectedText.append("@Override\n");
+		}
+		expectedText.append("public void run() {\n\t");
+		if (supportSnippets) {
+			expectedText.append("${0}");
+		}
+		expectedText.append("\n}");
+
+		assertEquals(expectedText.toString(), text);
+
```

Comment:
It would be cool if you could apply the completion item against the document and have an assert for the full content. Right now its not so easy to see if the completion proposal really does the right thing (e.g. in respect of indentation, or imports added)